pkg/volume/iscsi/iscsi_util.go

```diff
@@ -553,10 +553,18 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
 		glog.Warningf("Warning: Unmount skipped because path does not exist: %v", mntPath)
 		return nil
 	}
-	if err := c.mounter.Unmount(mntPath); err != nil {
-		glog.Errorf("iscsi detach disk: failed to unmount: %s\nError: %v", mntPath, err)
+
+	notMnt, err := c.mounter.IsLikelyNotMountPoint(mntPath)
+	if err != nil {
 		return err
 	}
+	if !notMnt {
```

Comment:
Kubelet itself makes sure that only one `UnmountDevice` runs for a given volume. So it can't change unless something else (admin) unmounts things in parallel.

And even if the volume is unmounted by admin after the check, the volume plugin calls `umount`, which fails, and kubelet re-tries `UnmountDevice` in a short while, now with correct check.

Reply:
It just feels akin to file exist check before deleting it, just deleting is more correct and no races